Generate client using OpenAPI Spec, build project, and update Package.md

linukaratnayake opened the pull request:
## Purpose

## Examples

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

linukaratnayake committed:
Correct module version to 1.0.0

linukaratnayake committed:
Add generated files

linukaratnayake committed:
Change Java version to 17

linukaratnayake committed:
Add Dependencies.toml file

linukaratnayake committed:
Add OpenAPI Spec file

linukaratnayake committed:
Add documentation to a record to avoid warnings

linukaratnayake committed:
Update 'Overview' section of Package.md

linukaratnayake committed:
Update 'Setup Guide' section of Package.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ballerina-platform/module-ballerinax-hubspot.crm.engagement.notes?pullRequest=2) <br/>All committers have signed the CLA.

linukaratnayake committed:
Modify OpenAPI spec and generate client

- Modify OpenAPI specification file by extracting base url from the endpoints and adding to server url.
- Generate new ballerina files using OpenAPI tool.
- Add documentation to two missing lines in auto-generated 'types.bal' file.

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
Write tests for create, update, delete, search notes

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
Write tests for batch create, update, delete notes

linukaratnayake committed:
Update sanitations.md

linukaratnayake committed:
Merge pull request #1 from linukaratnayake/openapi_sanitation

Perform sanitation on OpenAPI Spec

linukaratnayake committed:
Add license header to generated files

linukaratnayake committed:
Update documentation files

linukaratnayake committed:
Complete writing tests for live server

linukaratnayake committed:
Merge pull request #2 from linukaratnayake/tests

Tests

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
Format client.bal

ayeshLK commented on README.md:
```diff
@@ -8,11 +8,120 @@
+
+5. Store the access token securely for use in your application.
 
 ## Quickstart
```
Can you update the `Quickstart` section as well.

ayeshLK commented on ballerina/client.bal:
```diff
@@ -1,3 +1,6 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
+
 // Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
```
Check for other places as well.
```suggestion
// Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
```

ayeshLK commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,293 @@
+    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
+};
+
+// configurable http:BearerTokenConfig & readonly authConfig = ?;
+// ConnectionConfig config = {auth : authConfig};
+
+ConnectionConfig config = {auth: auth};
+
+final Client baseClient = check new Client(config, serviceUrl = "https://api.hubapi.com/crm/v3/objects/notes");
```
```suggestion
final Client hubSpotNotes = check new ({ auth });
```

ayeshLK commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,293 @@
+
+@test:Config {}
+isolated function testPost_batch_upsert_upsert() returns error? {
+    // BatchInputSimplePublicObjectBatchInputUpsert payload =
```
Are we using this test case ? If not lets remove it.

ayeshLK commented on ballerina/types.bal:
```diff
@@ -0,0 +1,368 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
+
+// Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
```
```suggestion
// Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
```

linukaratnayake committed:
Update documentation files with correct markdown format

ayeshLK commented on docs/spec/sanitations.md:
````diff
@@ -1,24 +1,33 @@
 
 ```bash
-# TODO: Add OpenAPI CLI command used to generate the client
+bal openapi -i docs/spec/notes.json --mode client  --license docs/license.txt -o ballerina
````
Why have not we committed the `license.txt` file ?

linukaratnayake replied:
The question is not clear. Can you explain what's wrong?

ayeshLK replied:
In the OpenAPI command you have mentioned that we need to use `license.txt` file. But, it has not been committed to the repository.

linukaratnayake replied:
I think `license.txt` file has been there in the `docs` directory since the beginning. Can you please check again?

Additionally, that file contains "2024" as the year. Should we change it to "2025" as well?

ayeshLK replied:
> I think `license.txt` file has been there in the `docs` directory since the beginning. Can you please check again?
> 
> Additionally, that file contains "2024" as the year. Should we change it to "2025" as well?

Agreed. Yes please change the year.

linukaratnayake committed:
Apply suggestions from code review

- Change year in header from 2024 to 2025.
- Apply suggestion in tests.bal

Co-authored-by: Ayesh Almeida <77491511+ayeshLK@users.noreply.github.com>

linukaratnayake committed:
Make some suggested changes

linukaratnayake committed:
Create mock tests

- Remove 'batch/upsert' test which does not work in the live server.
- Create mock service and client to test 'batch/upsert'.

linukaratnayake committed:
Create mock tests

- Remove 'batch/upsert' test which does not work in the live server.
- Create mock service and client to test 'batch/upsert'.

ayeshLK commented on ballerina/tests/mock_test.bal:
We don't need another set of tests as mock-tests. We should be able to run the original test suite in the `test.bal` file against the mock server. You can use Ballerina twitter connector as a reference on how to do this:

https://github.com/ballerina-platform/module-ballerinax-twitter/tree/main/ballerina/tests

linukaratnayake replied:
Two sets of tests address different endpoints. 10 tests in the `test.bal` call the live server, and 1 test (upsert) is mocked using the mock server. None of the tests use both live server and mock service. Hence, the tests are not written twice.

@ThisaruGuruge commented this is okay.

ayeshLK replied:
But we should be able to do this within single test suite right (by using test-groups). Please refer to Ballerina Twitter connector on how to do that.

https://github.com/ballerina-platform/module-ballerinax-twitter/tree/main/ballerina/tests

ayeshLK commented on docs/spec/notes.json:
Lets rename this file as `openapi.json`

ayeshLK commented on gradle.properties:
```diff
@@ -1,6 +1,6 @@
 org.gradle.caching=true
 group=io.ballerina.lib
-version=0.1.0-SNAPSHOT
+version=1.0.0
```
```suggestion
version=1.0.0-SNAPSHOT
```

linukaratnayake committed:
Rename OpenAPI Spec file

linukaratnayake committed:
Update gradle.properties

Co-authored-by: Ayesh Almeida <77491511+ayeshLK@users.noreply.github.com>

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
Create one example for operations on a single note

linukaratnayake committed:
Create one example for operations on a batch of notes

linukaratnayake committed:
Update .md files for examples

linukaratnayake committed:
Update image URLs and add example section to .md files

linukaratnayake committed:
Add 'Quick Start' section to .md files

linukaratnayake committed:
Change year from 2024 to 2025 in license file

linukaratnayake committed:
[Automated] Update the toml files

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+
+```ballerina
+import ballerina/oauth2;
+import ballerinax/hubspot.crm.engagement.notes as engagementNotes;
````
```suggestion
import ballerinax/hubspot.crm.engagement.notes as hsengnotes;
```

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+
+### Step 2: Instantiate a new connector
+
+1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:
+
+   ```toml
+   clientId = "<Client Id>"
+   clientSecret = "<Client Secret>"
+   refreshToken = "<Refresh Token>"
+   ```
+
+2. Instantiate `engagementNotes:OAuth2RefreshTokenGrantConfig` with the obtained credentials and initialize the connector with it.
````
Lets switch the order.

linukaratnayake replied:
Can you elaborate more on this suggestion?

ayeshLK replied:
Let's switch the order of point 1 and 2.

linukaratnayake replied:
Okay.

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+
+### Step 2: Instantiate a new connector
+
+1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:
+
+   ```toml
+   clientId = "<Client Id>"
+   clientSecret = "<Client Secret>"
+   refreshToken = "<Refresh Token>"
+   ```
+
+2. Instantiate `engagementNotes:OAuth2RefreshTokenGrantConfig` with the obtained credentials and initialize the connector with it.
````
```suggestion
1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
   clientId = "<Client Id>"
   clientSecret = "<Client Secret>"
   refreshToken = "<Refresh Token>"
   ```

2. Instantiate `hsengnotes:OAuth2RefreshTokenGrantConfig` with the obtained credentials and initialize the connector with it.
```

ayeshLK commented on README.md:
```diff
@@ -8,21 +8,173 @@
+    configurable string clientSecret = ?;
+    configurable string refreshToken = ?;
+
+    engagementNotes:OAuth2RefreshTokenGrantConfig auth = {
+        clientId,
+        clientSecret,
+        refreshToken,
+        credentialBearer: oauth2:POST_BODY_BEARER
+    };
+
+    final engagementNotes:Client hubSpotNotes = check new ({auth});
```
```suggestion
    hsengnotes:OAuth2RefreshTokenGrantConfig auth = {
        clientId,
        clientSecret,
        refreshToken,
        credentialBearer: oauth2:POST_BODY_BEARER
    };

    final hsengnotes:Client hubSpotNotes = check new ({ auth });
```

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+#### Get the note with a given ID
+
+```ballerina
+public function main() returns error? {
+string noteId = ""; // ID of the note that needs to be read
+engagementNotes:SimplePublicObjectWithAssociations readResponse = check hubSpotNotes->/[noteId]();
+}  
````
```suggestion
public function main() returns error? {
string noteId = ""; // ID of the note that needs to be read
hsengnotes:SimplePublicObjectWithAssociations readResponse = check hubSpotNotes->/[noteId]();
}  
```

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+#### Get the note with a given ID
+
+```ballerina
+public function main() returns error? {
````
We need to properly format this code section.

ayeshLK commented on README.md:
````diff
@@ -8,21 +8,173 @@
+engagementNotes:SimplePublicObjectWithAssociations readResponse = check hubSpotNotes->/[noteId]();
+}  
+```
 
````
We need to add a section on `Running the application` to `Quickstart` guide. Please refer to the following:

https://github.com/ballerina-platform/module-ballerinax-twitter/blob/main/README.md#step-4-run-the-ballerina-application

ayeshLK commented on examples/manage_notes/.devcontainer.json:
Remove this file.

ayeshLK commented on examples/manage_notes/.gitignore:
We can remove this as we already have a `.gitignore` file in the root of the repository.

ayeshLK commented on examples/manage_notes/Ballerina.toml:
```diff
@@ -0,0 +1,14 @@
+
+[build-options]
+observabilityIncluded = true
+
+[[dependency]]
+org= "ballerinax"
+name = "hubspot.crm.engagement.notes"
+version = "1.0.0"
+repository = "local"
```
```suggestion
```

ayeshLK commented on examples/manage_notes/main.bal:
```diff
@@ -0,0 +1,105 @@
+import ballerina/http;
```
Let's add WSO2 licensing header.

ayeshLK commented on examples/manage_notes/main.bal:
```diff
@@ -0,0 +1,105 @@
+import ballerina/io;
+import ballerina/oauth2;
+import ballerina/time;
+import ballerinax/hubspot.crm.engagement.notes as engagementNotes;
```
```suggestion
import ballerinax/hubspot.crm.engagement.notes as hsengnotes;
```

ayeshLK commented on examples/manage_notes/main.bal:
```diff
@@ -0,0 +1,105 @@
+// ID of the test company created for testing
+configurable string companyId = ?;
+
+engagementNotes:OAuth2RefreshTokenGrantConfig auth = {
+    clientId,
+    clientSecret,
+    refreshToken,
+    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
+};
+
+final engagementNotes:Client hubSpotNotes = check new ({auth});
```
```suggestion
hsengnotes:OAuth2RefreshTokenGrantConfig auth = {
    clientId,
    clientSecret,
    refreshToken,
    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
};
final hsengnotes:Client hubSpotNotes = check new ({ auth });
```

ayeshLK commented on examples/manage_notes_batch/.devcontainer.json:
Remove this file.

ayeshLK commented on examples/manage_notes_batch/.gitignore:
We can remove this as we already have a `.gitignore` at the root of the repository.

ayeshLK commented on examples/manage_notes_batch/Ballerina.toml:
```diff
@@ -0,0 +1,14 @@
+
+[build-options]
+observabilityIncluded = true
+
+[[dependency]]
+org= "ballerinax"
+name = "hubspot.crm.engagement.notes"
+version = "1.0.0"
+repository = "local"
```
```suggestion
```

ayeshLK commented on examples/manage_notes_batch/main.bal:
```diff
@@ -0,0 +1,128 @@
+import ballerina/http;
```
Add the WSO2 licensing header.

ayeshLK commented on examples/manage_notes_batch/main.bal:
```diff
@@ -0,0 +1,128 @@
+// ID of the test company created for testing
+configurable string companyId = ?;
+
+engagementNotes:OAuth2RefreshTokenGrantConfig auth = {
+    clientId,
+    clientSecret,
+    refreshToken,
+    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
+};
+
+final engagementNotes:Client hubSpotNotes = check new ({auth});
```
```suggestion
hsengnotes:OAuth2RefreshTokenGrantConfig auth = {
    clientId,
    clientSecret,
    refreshToken,
    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
};
final hsengnotes:Client hubSpotNotes = check new ({ auth });
```

ayeshLK commented on examples/manage_notes_batch/main.bal:
```diff
@@ -0,0 +1,128 @@
+import ballerina/io;
+import ballerina/oauth2;
+import ballerina/time;
+import ballerinax/hubspot.crm.engagement.notes as engagementNotes;
```
```suggestion
import ballerinax/hubspot.crm.engagement.notes as hsengnotes;
```

linukaratnayake committed:
Apply suggestions from code review

Co-authored-by: Ayesh Almeida <77491511+ayeshLK@users.noreply.github.com>

linukaratnayake committed:
Merge branch 'main' of https://github.com/linukaratnayake/module-ballerinax-hubspot.crm.engagement.notes

linukaratnayake committed:
Update .md files

linukaratnayake committed:
Add suggestions regarding examples

linukaratnayake committed:
Add test groups

linukaratnayake committed:
Add support to get environment variables for tests

ayeshLK commented on ballerina/tests/test.bal:
Can you restructure your test cases following the below structure ?

https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/blob/427167abf43333bfbf35c7c166f6329f8a4ea485/ballerina/tests/tests.bal

And if you want to run a particular test case only with the live-server, then lets update the `test:Config` annotation like below:

```ballerina
@test:Config {
     groups: ["live_tests"],
     enabled: isLiveServer
 }
```

ayeshLK commented on build-config/resources/Ballerina.toml:
```diff
@@ -5,12 +5,12 @@ name = "hubspot.crm.engagement.notes"
 authors = ["Ballerina"]
-keywords = [] # TODO: Add keywords
+keywords = ["hubspot", "crm", "engagement", "notes"]
 # icon = "icon.png" # TODO: Add icon
```
I think you have missed to add an `icon.png` file to the `ballerina` directory. Please add it and uncomment this line.

ayeshLK commented on examples/manage_notes/main.bal:
```diff
@@ -0,0 +1,123 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
+
```
Example are not auto-generated. Hence, lets remove this.
```suggestion
```

linukaratnayake committed:
Update .md files related to examples

ayeshLK commented on examples/manage_notes_batch/main.bal:
```diff
@@ -0,0 +1,146 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
+
```
```suggestion
```

ayeshLK commented on examples/manage_notes/manage_notes.md:
Lets rename this file to `Manage Notes.md`

linukaratnayake replied:
What if I rename it to `README.md`? Otherwise it will not be displayed automatically in the GitHub repository.

ayeshLK replied:
There is a trick to make it work [1] This naming convention is required for the SEO purposes.

[1] - https://stackoverflow.com/questions/49980639/specify-alternate-project-level-readme-md-on-github/49981731#49981731

ayeshLK commented on examples/manage_notes_batch/manage_notes_batch.md:
Lets rename this file to `Manage Notes Batch.md`

linukaratnayake replied:
What if I rename it to `README.md`? Otherwise it will not be displayed automatically in the GitHub repository.

ayeshLK replied:
Check this comment:

https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.engagement.notes/pull/2#discussion_r1909944698

linukaratnayake committed:
Rename .md files to README.md to display properly

linukaratnayake committed:
Add icon

linukaratnayake committed:
Apply suggestions from code review

Co-authored-by: Ayesh Almeida <77491511+ayeshLK@users.noreply.github.com>

linukaratnayake committed:
Switch order of 'Quick Start' points in .md files

linukaratnayake committed:
Change file to select when to run live and mock tests

linukaratnayake committed:
[Automated] Update the toml files

linukaratnayake committed:
Rename .md files in examples with their respective names

linukaratnayake committed:
Change year from 2024 to 2025

RavinduWeerakoon commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,323 @@
+// ID of the test company created for testing
+configurable string companyId = ?;
+
+OAuth2RefreshTokenGrantConfig auth = {
```
```suggestion
OAuth2RefreshTokenGrantConfig auth = {
```

linukaratnayake committed:
Add absolute path in README.md in examples

linukaratnayake committed:
Add function to initialize client

ayeshLK commented on README.md:
```diff
@@ -8,21 +8,179 @@
+
+If you don't have an account, you can sign up to a free account [here](https://developers.hubspot.com/get-started).
+
+### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account
```
During the review it was discussed to move the link to the description. Please fix the other places as well.

ayeshLK commented on examples/manage_notes_batch/main.bal:
```diff
@@ -0,0 +1,143 @@
+    clientId,
+    clientSecret,
+    refreshToken,
+    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
```
Lets remove the unwanted comments.
```suggestion
    credentialBearer: oauth2:POST_BODY_BEARER
```

ayeshLK commented on examples/manage_notes/main.bal:
```diff
@@ -0,0 +1,120 @@
+    clientId,
+    clientSecret,
+    refreshToken,
+    credentialBearer: oauth2:POST_BODY_BEARER // This line should be added when you are going to create auth object.
```
```suggestion
    credentialBearer: oauth2:POST_BODY_BEARER
```

linukaratnayake committed:
Apply suggestions from code review

Co-authored-by: Ayesh Almeida <77491511+ayeshLK@users.noreply.github.com>

linukaratnayake committed:
Move link from title to body in .md files

linukaratnayake committed:
Modify tests to run without Config.toml file
